packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -944,26 +924,50 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 	}
 
 	public async getPendingBlobs(waitBlobsToAttach?: boolean): Promise<IPendingBlobs | undefined> {
-		if (waitBlobsToAttach) {
-			await this.stashPendingBlobs();
-		}
-
-		if (this.pendingBlobs.size === 0) {
-			return;
-		}
-
-		const blobs = {};
-		for (const [key, entry] of this.pendingBlobs) {
-			blobs[key] = {
-				blob: bufferToString(entry.blob, "base64"),
-				storageId: entry.storageId,
-				attached: entry.attached,
-				acked: entry.acked,
-				minTTLInSeconds: entry.minTTLInSeconds,
-				uploadTime: entry.uploadTime,
-			};
-		}
-		return blobs;
+		return PerformanceEvent.timedExecAsync(
+			this.mc.logger,
+			{ eventName: "GetPendingBlobs" },
+			async () => {
+				if (this.pendingBlobs.size === 0) {
+					return;
+				}
+				const blobs = {};
+				const localIds = new Set<PendingBlob>();
+				while (localIds.size < this.pendingBlobs.size) {
+					for (const [id, entry] of this.pendingBlobs) {
+						if (!localIds.has(entry)) {
+							localIds.add(entry);
+							if (waitBlobsToAttach) {
+								if (!entry.opsent) {
+									this.sendBlobAttachOp(id, entry.storageId);
+								}
+								entry.handleP.resolve(this.getBlobHandle(id));
+								if (!entry.attached) {
+									await new Promise<void>((resolve) => {
```

Comment:
rather than await here, it would be better to collect all the promises, and wait on them all at once with Promise.all, as this will force the blobs to be handled one by one, and could lead to a deadlock if a consumer needs mutliple blob handles for some reason